Add in initial parser and transpiler code

ameamenoame opened the pull request:
Add in the very messy code for the parser and the transpiler.

Related issue (if applicable): #10

ameamenoame committed:
Add in initial parser and transpiler code

ameamenoame committed:
Properly handle choice trees in the transpiler

ameamenoame committed:
Almost feature-complete parser

ameamenoame committed:
Add support for elif blocks

ameamenoame committed:
Minor refactoring in the transpiler

ameamenoame committed:
Support a basic `scene` command

ameamenoame committed:
Clean up, add more comments and rename variables

ameamenoame committed:
Add error push for the yet unsupported `pass` command

ameamenoame committed:
Fix if statements without else breaking

ameamenoame committed:
Fix inner code blocks not continuing on properly to their parent blocks

ameamenoame committed:
Allow `mark` commands to work inside code blocks

ameamenoame committed:
Remove unnecessary code and change some comments

ameamenoame committed:
Add some variable declarations to the test-scene script

NathanLovato commented on godot/Parser/SceneTranspiler.gd:
```diff
@@ -0,0 +1,449 @@
+	const GLOBALS := {
+		jump_points = {},
+		variables = {}
+		}
```
If this is supposed to be available across the game, so shared across all scenes, you shouldn't define the constant here. It's never going to be a global either so I'd use two variables, say, on the main class, or on some singleton. But not on what's supposed to represent a single dialogue tree.

ameamenoame replied:
I've moved the `jump_points` somewhere else and used a singleton for handling variables.

NathanLovato reviewed (changes requested):
There's just one change I'd like with the GLOBALS property. You can remove it and store `variables` and `jump_points` in an autoload if that's convenient, otherwise in like `Main.gd` and pass it to `ScenePlayer` instances.

Best not present them as constants here as they're not constant (variables can be any value and change anytime with the player's decisions, jump points get updated with parsed scenes).

I'll change other things myself for the tutorials, like breaking some code into functions, but that's beyond the scope of your job. You did some great work already.

ameamenoame committed:
Remove GLOBALS from the DialogueTree class

Jump points are stored per script. Consolidate the logic for handling global variables
into a Variables singleton.

NathanLovato committed:
Move constant to the top of SceneParser

NathanLovato committed:
Reorder code, add docstrings in SceneTranspiler

NathanLovato committed:
Refactor SceneTranspiler's command conversion, rename a property

NathanLovato committed:
Refactor SceneTranspiler

Extract transpile dialogue code to a function
Rename symbols
Add a constant for -3

NathanLovato commented:
Great job! There are some things left to tweak and refactor but it's time to merge already.
